hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java

```diff
@@ -242,7 +253,16 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
         if (null == oldKeyVersionsToDeleteMap) {
           oldKeyVersionsToDeleteMap = new HashMap<>();
         }
-        oldKeyVersionsToDeleteMap.put(delKeyName, oldVerKeyInfo);
+
+        // Remove any block from oldVerKeyInfo that share the same container ID
+        // and local ID with omKeyInfo blocks'.
```

Comment:
The fix is good to avoid deletion of blocks in generic. I think actual root cause can also be fixed. Attached fix for this.

[hsync_delete_issue.patch](https://github.com/apache/ozone/files/12312760/hsync_delete_issue.patch)

Reply:
Thanks @sumitagrawl . Similar idea also occurred to us but I wasn't proficient in HSync block commit flow so @errose28 , @jojochuang and I decided to take the generic filtering approach, for now.

1. I briefly tried out your patch on top of this PR branch. I get the idea. However, if you comment out the call to the filtering logic:

https://github.com/apache/ozone/blob/5661ad98616245606baffd09c86472d6af9c2968/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java#L214

and run the new integration test, you'll see `OMKeyCommitRequestWithFSO` is still putting the key block into `deletedTable`:

```
org.opentest4j.AssertionFailedError: deletedTable should not have such entry. key = /volume13960/bucket38943/file-hsync-then-close/4611686018427388673

	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
	at org.junit.jupiter.api.Assertions.fail(Assertions.java:134)
	at org.apache.hadoop.fs.ozone.TestHSync.testKeyHSyncThenClose(TestHSync.java:184)
```

That is because when it is reading the `omKeyInfo` during the second non-HSync commit, `omKeyInfo.getMetadata()` is empty. So `isPreviousCommitHsync` is always `false` (at least in FSO).

In turn this is because `omKeyInfo` is retrieved from `OpenKeyTable`:

https://github.com/apache/ozone/blob/5661ad98616245606baffd09c86472d6af9c2968/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java#L153-L154

and in KeyCommitRequest we only put the metadata in `KeyTable`, not `OpenKeyTable`:

https://github.com/apache/ozone/blob/5661ad98616245606baffd09c86472d6af9c2968/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestWithFSO.java#L252-L253

Therefore, the patch itself it looks incomplete. Would you raise a another PR once this filtering patch is merged?

2. Another problem is, currently I see no check to ensure the same `clientId` is doing the commit in HSync key commit flow. Therefore, if another client with different ClientID (rogue or not) tries to commit this key without HSync, `isPreviousCommitHsync` will be `false` according to your patch. Then again it will fall into the second branch, which queues the block for deletion and in turn cause data loss if the block filtering logic does not exist.

Reply:
All in all, I think we will need to further:

1) Update OpenKeyTable with the HSync metadata as well? (and check if such action would lead to further unintended side effects)
2) Improve or fix the clientId check
3) Sort out the HSync and non-HSync key commit flow to make sure no other code paths would lead to unintended consequences like this. And fix them if we found any other issues.

in a series of follow-up patches.

Reply:
isPreviousCommitHsync needs to be true whenever HSYNC_CLIENT_ID has a value.

Reply:
Hsync flow: openKeyTable/openFileTable will have hsync flag whenever first commit comes with hsync feature
1. openKeyTable: when first commit comes with hsync flag true (as hsync feature), it will add to metadata of open key table
2. Multiple commits can come with hsync flag true (always)
3. When final commit comes (with hsync flag false), it will retrive same flag from metadata of open key table

Related to ClientId,
- commit always retrieve omKeyInfo from with clientId as full keyName, else commit will fail. So there can not be any other annonymous clientId
`String dbOpenKey = omMetadataManager.getOpenKey(volumeName, bucketName,
          keyName, commitKeyRequest.getClientID());`

We can take up this as separate verifying if any other issue, and this PR looks good to me.